common/src/main/java/org/apache/celeborn/common/network/ssl/ReloadingX509TrustManager.java

```diff
@@ -0,0 +1,218 @@
+    this.reloadCount = 0;
+    this.needsReloadCheckCounts = 0;
+  }
+
+  /** Starts the reloader thread. */
+  public void init() {
+    reloader = new Thread(this, "Truststore reloader thread");
+    reloader.setDaemon(true);
+    reloader.start();
+  }
+
+  /** Stops the reloader thread. */
+  public void destroy() throws InterruptedException {
+    reloader.interrupt();
+    reloader.join();
+  }
+
+  /**
+   * Returns the reload check interval.
+   *
+   * @return the reload check interval, in milliseconds.
+   */
+  public long getReloadInterval() {
+    return reloadInterval;
+  }
+
+  @Override
+  public void checkClientTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    X509TrustManager tm = trustManagerRef.get();
+    if (tm != null) {
+      tm.checkClientTrusted(chain, authType);
+    } else {
+      throw new CertificateException(
+          "Unknown client chain certificate: "
+              + chain[0].toString()
+              + ". Please ensure the correct trust store is specified in the config");
+    }
+  }
+
+  @Override
+  public void checkServerTrusted(X509Certificate[] chain, String authType)
+      throws CertificateException {
+    X509TrustManager tm = trustManagerRef.get();
+    if (tm != null) {
+      tm.checkServerTrusted(chain, authType);
+    } else {
+      throw new CertificateException(
+          "Unknown server chain certificate: "
+              + chain[0].toString()
+              + ". Please ensure the correct trust store is specified in the config");
+    }
+  }
+
+  private static final X509Certificate[] EMPTY = new X509Certificate[0];
+
+  @Override
+  public X509Certificate[] getAcceptedIssuers() {
+    X509Certificate[] issuers = EMPTY;
+    X509TrustManager tm = trustManagerRef.get();
+    if (tm != null) {
+      issuers = tm.getAcceptedIssuers();
+    }
+    return issuers;
+  }
+
+  boolean needsReload() throws IOException {
+    boolean reload = true;
+    File latestCanonicalFile = file.getCanonicalFile();
+    if (file.exists() && latestCanonicalFile.exists()) {
+      // `file` can be a symbolic link. We need to reload if it points to another file,
+      // or if the file has been modified
+      if (latestCanonicalFile.getPath().equals(canonicalPath)
+          && latestCanonicalFile.lastModified() == lastLoaded) {
+        reload = false;
+      }
+    } else {
+      lastLoaded = 0;
+    }
+    return reload;
+  }
+
+  X509TrustManager loadTrustManager() throws IOException, GeneralSecurityException {
+    X509TrustManager trustManager = null;
+    KeyStore ks = KeyStore.getInstance(type);
+    File latestCanonicalFile = file.getCanonicalFile();
+    canonicalPath = latestCanonicalFile.getPath();
+    lastLoaded = latestCanonicalFile.lastModified();
+    try (FileInputStream in = new FileInputStream(latestCanonicalFile)) {
+      char[] passwordCharacters = password != null ? password.toCharArray() : null;
+      ks.load(in, passwordCharacters);
```

Comment:
Can concurrent read/write happen here?

Reply:
`load()` returns a new instance of `X509TrustManager` if the file is detected to be modified - which is saved in the `trustManagerRef` atomic reference - so it is thread safe from point of view of use of the trust manager itself.

Or did you mean, concurrent writes to the file while there are ongoing reads ? If yes, we expect this to be atomically renamed - create new file in the same directory, and then execute an `mv`.

Reply:
While this is not there in spark docs, for Celeborn we will add an additional note on expectation on reload.
This is going to be applicable for the server side primarily - not application side; and we would want to make sure any update is propagated consistently in the fleet.

So `trustStoreReloadingEnabled` would say something like - make sure you atomically update the truststore by copying to the directory and doing an atomic move (using `mv`)

Reply:
Thanks for the explanation, I do mean concurrent writes to the file while it's being read. I think `mv` can work in such case. Merging to main(v0.5.0)